Supported AWS Fargate

kakakakakku opened the pull request:
# WHAT IS THIS

AWS released new service "AWS Fargate" on November 30, 2017.

- https://aws.amazon.com/fargate/

This Pull Request goal is supporting AWS Fargate deployment, and solve this Issue ( https://github.com/silinternational/ecs-deploy/issues/127 ).

# ERROR

This error occurred by latest ecs-deploy.

```
An error occurred (InvalidParameterException) when calling the UpdateService operation: Task definition does not support launch_type FARGATE.
```

Because `Launch Types` ( parameter name is `requiresCompatibilities` ) added on Task Definition.

- https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task_definition_parameters.html

Then `requiresCompatibilities` set value `FARGATE` or `EC2`, but `EC2` is default ECS.

- https://docs.aws.amazon.com/AmazonECS/latest/developerguide/launch_types.html

# REQUIRE

If you deploy AWS Fargate using `aws-cli 1.14.0 or later` please.

# DETAILS

- Check `requiresCompatibilities` via Task Definition (JSON)
- If value equal `FARGATE` jq filters updated

Deployment is successful in my Fargate Cluster.

```
New task definition: arn:aws:ecs:us-east-1:111111111111:task-definition/xxx:13
Service updated successfully, new task definition running.
Waiting for service deployment to complete...
Service deployment successful.
```

# ANY HELP

I am little worried about deployment for other Fargate Cluster.

If someone has own Fargate Cluster, please check deployment with my version ??? 🙏 
~~( This is reason why this Pull Request is [WIP] . )~~

Thanks 👍 👍 👍

tacchino commented:
Worked perfectly for me

kakakakakku commented:
@tacchino 
Wow! Thank you for your testing 👍

zeroasterisk commented:
@kakakakakku it is not working for me yet, sadly

```
$ ecs-deploy -c portal -n "etx-onspot-prod" -i "000000000000.dkr.ecr.us-east-1.amazonaws.com/etx-onspot:latest"
Using image name: 000000000000.dkr.ecr.us-east-1.amazonaws.com/etx-onspot:latest
Current task definition: arn:aws:ecs:us-east-1:000000000000:task-definition/etx-onspot-prod:1
jq: error (at <stdin>:131): Cannot iterate over null (null)
```

Here's a verbose dump, with env. vars truncated and account id redacted
http://sebsauvage.net/paste/?3f76a75b911230c8#FDlFZAxjQJwA1oAPflGIrsogE6Ll/A10wAFodUpNPcs=

fillup commented:
@kakakakakku thank you for your work on this. I am getting the same error as @zeroasterisk. I upgraded my local awscli to 1.14.10 in case it was a change to the API but that didn't help. For reference I am testing updating a task/service that is NOT using Fargate. 

Also unit tests fail. If you have Docker and Docker Compose installed you can run `docker-compose run --rm test` to work through tests and updating them. Here is the output from me running the tests right now:

```
$ docker-compose run --rm test
fetch http://dl-3.alpinelinux.org/alpine/edge/community/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.5/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.5/community/x86_64/APKINDEX.tar.gz
(1/1) Installing bats (0.4.0-r2)
Executing busybox-1.25.1-r0.trigger
OK: 64 MiB in 31 packages
1..32
ok 1 # skip (0) check that usage() returns string and exits with status code 20
ok 2 # skip (0) test assertRequiredArgumentsSet success
ok 3 # skip (0) test assertRequiredArgumentsSet status=5
ok 4 # skip (0) test assertRequiredArgumentsSet status=6
ok 5 # skip (0) test assertRequiredArgumentsSet status=7
ok 6 # skip (0) test assertRequiredArgumentsSet status=8
ok 7 # skip (0) test assertRequiredArgumentsSet status=9
ok 8 # skip (0) test parseImageName missing image name
ok 9 # skip (0) test parseImageName invalid image name 1
ok 10 # skip (0) test parseImageName invalid port
ok 11 # skip (0) test parseImageName root image no tag
ok 12 # skip (0) test parseImageName root image with tag
ok 13 # skip (0) test parseImageName repo image no tag
ok 14 # skip (0) test parseImageName repo image with tag
ok 15 # skip (0) test parseImageName repo multilevel image no tag
ok 16 # skip (0) test parseImageName repo multilevel image with tag
ok 17 # skip (0) test parseImageName domain plus repo image no tag
ok 18 # skip (0) test parseImageName domain plus repo image with tag
ok 19 # skip (0) test parseImageName domain plus repo multilevel image no tag
ok 20 # skip (0) test parseImageName domain plus repo multilevel image with tag
ok 21 # skip (0) test parseImageName domain plus port plus repo image no tag
ok 22 # skip (0) test parseImageName domain plus port plus repo image with tag
ok 23 # skip (0) test parseImageName domain plus port plus repo multilevel image no tag
ok 24 # skip (0) test parseImageName domain plus port plus repo multilevel image with tag
ok 25 # skip (0) test parseImageName domain plus port plus repo image with tag from var
ok 26 # skip (0) test parseImageName domain plus port plus repo multilevel image with tag from var
ok 27 # skip (0) test parseImageName using ecr style domain
ok 28 # skip (0) test parseImageName using ecr style image name and tag from var
not ok 29 test createNewTaskDefJson with single container in definition
# (in test file test.bats, line 281)
#   `[ $output == $expected ]' failed with status 2
# /tmp/bats.40.src: line 281: [: too many arguments
not ok 30 test createNewTaskDefJson with multiple containers in definition
# (in test file test.bats, line 365)
#   `[ $output == $expected ]' failed with status 2
# /tmp/bats.40.src: line 365: [: too many arguments
ok 31 # skip (0) test parseImageName with tagonly option
not ok 32 test createNewTaskDefJson with multiple containers in definition and replace only tags
# (in test file test.bats, line 463)
#   `[ $output == $expected ]' failed with status 2
# jq: error (at <stdin>:70): Cannot iterate over null (null) { "family": "app-task-def", "volumes": [], "containerDefinitions": [ { "environment": [ { "name": "KEY", "value": "value" } ], "name": "API", "links": [], "mountPoints": [], "image": "121212345678.dkr.ecr.us-east-1.amazonaws.com/acct/repo:newtag", "essential": true, "portMappings": [ { "protocol": "tcp", "containerPort": 80, "hostPort": 10080 } ], "entryPoint": [], "memory": 128, "command": [ "/data/run.sh" ], "cpu": 200, "volumesFrom": [] }, { "environment": [ { "name": "KEY", "value": "value" } ], "name": "cache", "links": [], "mountPoints": [], "image": "redis:newtag", "essential": true, "portMappings": [ { "protocol": "tcp", "containerPort": 6376, "hostPort": 10376 } ], "entryPoint": [], "memory": 128, "command": [ "/data/run.sh" ], "cpu": 200, "volumesFrom": [] } ], "placementConstraints": null, "networkMode": "bridge" }
# /tmp/bats.40.src: line 463: [: too many arguments
```

kakakakakku commented:
@zeroasterisk @fillup 
Thanks for your reports !
I will check error, and title changed to [WIP] 👍

kakakakakku commented:
@zeroasterisk @fillup

Your task definition (JSON) seem doesn't have `requiresCompatibilities` key.
Thank you for your information.

So, We must support these patterns.

- [1] `requiresCompatibilities` key doesn't exist
- [2] `requiresCompatibilities` key equals `null`
- [3] `requiresCompatibilities` key equals `EC2`
- [4] `requiresCompatibilities` key equals `FARGATE`

Then I fixed with using `jq select`.

# Before

```sh
# [1]
$ echo '{ "status": "ACTIVE" }' | jq -r '.requiresCompatibilities[]'
jq: error (at <stdin>:1): Cannot iterate over null (null)

# [2]
$ echo '{ "status": "ACTIVE", "requiresCompatibilities": null }' | jq -r '.requiresCompatibilities[]'
jq: error (at <stdin>:1): Cannot iterate over null (null)

# [3]
$ echo '{ "status": "ACTIVE", "requiresCompatibilities": [ "EC2" ] }' | jq -r '.requiresCompatibilities[]'
EC2

# [4]
$ echo '{ "status": "ACTIVE", "requiresCompatibilities": [ "FARGATE" ] }' | jq -r '.requiresCompatibilities[]'
FARGATE
```

# After

```sh
# [1]
$ echo '{ "status": "ACTIVE" }' | jq -r '. | select(.requiresCompatibilities != null) | .requiresCompatibilities[]'

# [2]
$ echo '{ "status": "ACTIVE", "requiresCompatibilities": null }' | jq -r '. | select(.requiresCompatibilities != null) | .requiresCompatibilities[]'

# [3]
$ echo '{ "status": "ACTIVE", "requiresCompatibilities": [ "EC2" ] }' | jq -r '. | select(.requiresCompatibilities != null) | .requiresCompatibilities[]'
EC2

# [4]
$ echo '{ "status": "ACTIVE", "requiresCompatibilities": [ "FARGATE" ] }' | jq -r '. | select(.requiresCompatibilities != null) | .requiresCompatibilities[]'
FARGATE
```

# tests

I added tests, and pass succsessflly.

```
$ docker-compose run --rm test
fetch http://dl-3.alpinelinux.org/alpine/edge/community/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.5/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.5/community/x86_64/APKINDEX.tar.gz
(1/1) Installing bats (0.4.0-r2)
Executing busybox-1.25.1-r1.trigger
OK: 64 MiB in 31 packages
1..33
ok 1 # skip (0) check that usage() returns string and exits with status code 20
ok 2 # skip (0) test assertRequiredArgumentsSet success
ok 3 # skip (0) test assertRequiredArgumentsSet status=5
ok 4 # skip (0) test assertRequiredArgumentsSet status=6
ok 5 # skip (0) test assertRequiredArgumentsSet status=7
ok 6 # skip (0) test assertRequiredArgumentsSet status=8
ok 7 # skip (0) test assertRequiredArgumentsSet status=9
ok 8 # skip (0) test parseImageName missing image name
ok 9 # skip (0) test parseImageName invalid image name 1
ok 10 # skip (0) test parseImageName invalid port
ok 11 # skip (0) test parseImageName root image no tag
ok 12 # skip (0) test parseImageName root image with tag
ok 13 # skip (0) test parseImageName repo image no tag
ok 14 # skip (0) test parseImageName repo image with tag
ok 15 # skip (0) test parseImageName repo multilevel image no tag
ok 16 # skip (0) test parseImageName repo multilevel image with tag
ok 17 # skip (0) test parseImageName domain plus repo image no tag
ok 18 # skip (0) test parseImageName domain plus repo image with tag
ok 19 # skip (0) test parseImageName domain plus repo multilevel image no tag
ok 20 # skip (0) test parseImageName domain plus repo multilevel image with tag
ok 21 # skip (0) test parseImageName domain plus port plus repo image no tag
ok 22 # skip (0) test parseImageName domain plus port plus repo image with tag
ok 23 # skip (0) test parseImageName domain plus port plus repo multilevel image no tag
ok 24 # skip (0) test parseImageName domain plus port plus repo multilevel image with tag
ok 25 # skip (0) test parseImageName domain plus port plus repo image with tag from var
ok 26 # skip (0) test parseImageName domain plus port plus repo multilevel image with tag from var
ok 27 # skip (0) test parseImageName using ecr style domain
ok 28 # skip (0) test parseImageName using ecr style image name and tag from var
ok 29 # skip (0) test createNewTaskDefJson with single container in definition
ok 30 # skip (0) test createNewTaskDefJson with single container in definition for AWS Fargate
ok 31 # skip (0) test createNewTaskDefJson with multiple containers in definition
ok 32 # skip (0) test parseImageName with tagonly option
ok 33 # skip (0) test createNewTaskDefJson with multiple containers in definition and replace only tags
```

Please review again 👍

zeroasterisk commented:
@kakakakakku I am looking at https://github.com/kakakakakku/ecs-deploy/tree/fargate-support and your last commit is `675e0e064bc20ad81634efdfe41d802fbe3c30e3` <-- did this change? force push?

I updated it and ran it and got:
```
An error occurred (ClientException) when calling the RegisterTaskDefinition operation: Invalid setting for container 'etx-onspot'. At least one of 'memory' or 'memoryReservation' must be specified.
```

kakakakakku commented:
@zeroasterisk 
Yes, many PRs was merged yesterday then I rebase master and push force.

>did this change? force push?

Thank you for your reports.
I have questions.

- Do you try EC2 mode or Fargate mode ?
- If you use ecs-deploy with master branch, how about occurring such error ?

If you try EC2 mode, `memory` or `memoryReservation` is necessary in task definition. 
Your task definition ( https://github.com/silinternational/ecs-deploy/pull/129#issuecomment-352487835 ) doesn't have `memory` or `memoryReservation`.

>For containers that will be part of a task using the EC2 launch type, you must specify a non-zero integer for one or both of memory or memoryReservation in container definitions. 
https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task_definition_parameters.html

Thanks 👍

richardbolt commented:
I found success or failure to be AWS CLI version dependent. Older aws cli versions didn't work, newer ones did. For example we had aws v0.11.39 on a build machine which failed to update a Fargate service, but updating aws to 0.14.9 worked just fine for the same service.

kakakakakku commented:
@richardbolt 
I forgot to write !
Added require version, Thanks 👍

kakakakakku commented:
@zeroasterisk 
OK, I will try reproducing my Fargate Cluster.

Please give me your task definition (JSON), or http://sebsauvage.net/paste/?3f76a75b911230c8#FDlFZAxjQJwA1oAPflGIrsogE6Ll/A10wAFodUpNPcs= is it ?

zeroasterisk commented:
@kakakakakku 
Correct - after I updated my AWS CLI with `$ pip install awscli --upgrade --user` your fork now works for me.
Also Correct - the link you sent is my (redacted) task definition/log.

kakakakakku commented:
@zeroasterisk
Nice, thank you for your review👍

tacchino commented:
Is this ready to go now?

kakakakakku commented:
Yes, I'm OK !
Please review :)

shimont commented:
I tried the code on our system and the deployment worked but did not perform a graceful blue-green deployment. The code killed the running revision immediately and ran the new revision successfully on ECS, but produced a script error:

`+echo 'Deploying service:XXX on to cluster:dev with image:XXX.dkr.ecr.us-east-1.amazonaws.com/XXX:dev'`
`Deploying service:XXX on to cluster:dev with image:XXX.dkr.ecr.us-east-1.amazonaws.com/XXX:dev`
`+./scripts/ecs-deploy.sh --cluster dev --service-name XXX --image XXX.dkr.ecr.us-east-1.amazonaws.com/XXX:dev`
`Using image name: XXX.dkr.ecr.us-east-1.amazonaws.com/XXX:dev`
`Current task definition: arn:aws:ecs:us-east-1:XXX:task-definition/XXX-dev:1`
`New task definition: arn:aws:ecs:us-east-1:XXX:task-definition/XXX-dev:2`
`ERROR: New task definition not running within 90 seconds`

kakakakakku commented:
@shimont 
Thank you for your reports.
Are you trying "Fargete" deployment right ?

Please give me `diff of task-definition/XXX-dev:1 and task-definition/XXX-dev:2`.
I want to check your diff of task definition.

shimont commented:
@kakakakakku after performing further tests I think that the problem was on my end. The task didnt register correctly via the ALB due to an error in the configuration of the health check endpoint. I've performed the test again and I can confirm that a full cycle of deployment to FARGATE is **working great**! The only thing I suggest people do is change the --timeout argument from 90sec to 360sec as FARGATE can take some time to deploy the new container

kakakakakku commented:
@shimont 
Nice deployment 🎉🎉🎉 

Yes, I think so too ! (My opinion) Because "Fargate" require `awsvpc` networking mode, so ENI has attached to all containers and some initial time needed.

>FARGATE can take some time to deploy the new container

Owner, please review continuously 👍 
Thank you.

mohuk commented:
any plans for merging this ?

shimont commented:
Could you please merge this?

bazaglia commented:
Tested and approved. +1 for merging this.

himberjack commented:
Working! Please merge

cameronk commented:
Bumping this, looking forward to a merge!

adambiggs commented:
I can also confirm that this fixes Fargate deploys.

Please merge! :octocat:

myoung34 commented:
Bumping, need this sooner than later =)

fillup commented:
Hi all, I'm really sorry for taking forever to catch up with this. I just pulled this down but one of the new tests is failing, @kakakakakku can you check this?

```
not ok 30 test createNewTaskDefJson with single container in definition for AWS Fargate
# (in test file test.bats, line 347)
#   `[ $output == $expected ]' failed
```

Also as an FYI, this does not work with tasks that were created a while ago (dont, maybe pre-fargate?). When running this update against a test service I have it throws a different error that seems to be from the aws-cli itself as it has python type errors:
```
./ecs-deploy -c staging1 -n ecs-deploy-testing -i silintl/http-to-https:latest
Using image name: silintl/http-to-https:latest
Current task definition: arn:aws:ecs:us-east-1:1234567890:task-definition/staging-ecs-deploy-testing:40

Parameter validation failed:
Invalid type for parameter placementConstraints, value: None, type: <type 'NoneType'>, valid types: <type 'list'>, <type 'tuple'>
```
I went into the AWS console and created a new revision without changing anything and re-ran ecs-deploy and it worked just fine. 

When we get the test issue resolved I'll be happy to get this merged in.

kakakakakku commented:
@fillup 
Thanks !

But all tests are passed on December, see my comment / https://github.com/silinternational/ecs-deploy/pull/129#issuecomment-352642007 . 
I will run again 👍

kakakakakku commented:
@fillup 
Oh, I understand.

on January `placementConstraints` parameter supported via https://github.com/silinternational/ecs-deploy/pull/114 .
My Pull-Request created on December, so some parameter diifecent.

```diff
$ diff -u output.json expected.json
--- output.json	2018-03-31 13:00:13.000000000 +0900
+++ expected.json	2018-03-31 13:00:23.000000000 +0900
@@ -22,7 +22,6 @@
 		"cpu": 200,
 		"volumesFrom": []
 	}],
-	"placementConstraints": null,
 	"networkMode": "awsvpc",
 	"executionRoleArn": "arn:aws:iam::121212345678:role/ecsTaskExecutionRole",
 	"requiresCompatibilities": ["FARGATE"],
```

So I rebased from master and fixed it ( https://github.com/silinternational/ecs-deploy/pull/129/commits/423b0038e57bab7190fbb18aa8fd1f70a128567f ) .

# tests

```sh
$ docker-compose run --rm test
fetch http://dl-3.alpinelinux.org/alpine/edge/community/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.5/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.5/community/x86_64/APKINDEX.tar.gz
(1/1) Installing bats (0.4.0-r2)
Executing busybox-1.25.1-r1.trigger
OK: 65 MiB in 31 packages
1..33
ok 1 # skip (0) check that usage() returns string and exits with status code 20
ok 2 # skip (0) test assertRequiredArgumentsSet success
ok 3 # skip (0) test assertRequiredArgumentsSet status=5
ok 4 # skip (0) test assertRequiredArgumentsSet status=6
ok 5 # skip (0) test assertRequiredArgumentsSet status=7
ok 6 # skip (0) test assertRequiredArgumentsSet status=8
ok 7 # skip (0) test assertRequiredArgumentsSet status=9
ok 8 # skip (0) test parseImageName missing image name
ok 9 # skip (0) test parseImageName invalid image name 1
ok 10 # skip (0) test parseImageName invalid port
ok 11 # skip (0) test parseImageName root image no tag
ok 12 # skip (0) test parseImageName root image with tag
ok 13 # skip (0) test parseImageName repo image no tag
ok 14 # skip (0) test parseImageName repo image with tag
ok 15 # skip (0) test parseImageName repo multilevel image no tag
ok 16 # skip (0) test parseImageName repo multilevel image with tag
ok 17 # skip (0) test parseImageName domain plus repo image no tag
ok 18 # skip (0) test parseImageName domain plus repo image with tag
ok 19 # skip (0) test parseImageName domain plus repo multilevel image no tag
ok 20 # skip (0) test parseImageName domain plus repo multilevel image with tag
ok 21 # skip (0) test parseImageName domain plus port plus repo image no tag
ok 22 # skip (0) test parseImageName domain plus port plus repo image with tag
ok 23 # skip (0) test parseImageName domain plus port plus repo multilevel image no tag
ok 24 # skip (0) test parseImageName domain plus port plus repo multilevel image with tag
ok 25 # skip (0) test parseImageName domain plus port plus repo image with tag from var
ok 26 # skip (0) test parseImageName domain plus port plus repo multilevel image with tag from var
ok 27 # skip (0) test parseImageName using ecr style domain
ok 28 # skip (0) test parseImageName using ecr style image name and tag from var
ok 29 # skip (0) test createNewTaskDefJson with single container in definition
ok 30 # skip (0) test createNewTaskDefJson with single container in definition for AWS Fargate
ok 31 # skip (0) test createNewTaskDefJson with multiple containers in definition
ok 32 # skip (0) test parseImageName with tagonly option
ok 33 # skip (0) test createNewTaskDefJson with multiple containers in definition and replace only tags
```

fillup commented:
Thanks @kakakakakku for getting that fixed, it all looks good to me so I've merged it in and will roll out a release shortly
